python/ray/serve/_private/proxy_router.py

```diff
@@ -0,0 +1,189 @@
+from abc import ABC, abstractmethod
+import logging
+from typing import Callable, Dict, List, Optional, Tuple
+
+from ray.serve._private.common import (
+    ApplicationName,
+    EndpointInfo,
+    EndpointTag,
+)
+from ray.serve._private.constants import (
+    SERVE_LOGGER_NAME,
+    RAY_SERVE_ENABLE_EXPERIMENTAL_STREAMING,
+)
+from ray.serve.handle import RayServeHandle
+
+logger = logging.getLogger(SERVE_LOGGER_NAME)
+
+
+class ProxyRouter(ABC):
+    """Router interface for the proxy to use."""
+
+    @abstractmethod
+    def update_routes(self, endpoints: Dict[EndpointTag, EndpointInfo]):
+        raise NotImplementedError
+
+    @abstractmethod
+    def match_route(
+        self, target_route: str
+    ) -> Optional[Tuple[str, RayServeHandle, str, bool]]:
+        raise NotImplementedError
+
+
+class LongestPrefixRouter(ProxyRouter):
+    """Router that performs longest prefix matches on incoming routes."""
+
+    def __init__(self, get_handle: Callable):
+        # Function to get a handle given a name. Used to mock for testing.
+        self._get_handle = get_handle
+        # Routes sorted in order of decreasing length.
+        self.sorted_routes: List[str] = list()
+        # Endpoints associated with the routes.
+        self.route_info: Dict[str, EndpointTag] = dict()
+        # Contains a ServeHandle for each endpoint.
+        self.handles: Dict[EndpointTag, RayServeHandle] = dict()
+        # Map of application name to is_cross_language.
+        self.app_to_is_cross_language: Dict[ApplicationName, bool] = dict()
+
+    def endpoint_exists(self, endpoint: EndpointTag) -> bool:
```

Comment:
is this function being used anywhere?

Reply:
Oh good point, this whole class is just moved into this new file and added a base class. I did a code search and doesn't seem like it's being used currently. Will just drop it